### ISSUE TITLE
timeshift-spy@nicog60 - crashes with non default backup location

### ISSUE THREAD

**Issue by smurphos** [linuxmint/cinnamon-spices-applets#2998]
```
 * Cinnamon version 4.4.8
 * Distribution - Mint 19.3
 * Snippet or attachment of ~/.cinnamon/glass.log
 * Snippet or attachment of ~/.xsession-errors
(cinnamon:1856): Cjs-WARNING **: 08:04:51.056: JS ERROR: Gio.IOErrorEnum: Error opening directory '/timeshift/snapshots': No such file or directory
check_if_snapping@/home/steve/.local/share/cinnamon/applets/timeshift-spy@nicog60/applet.js:177:17
setInterval/id<@/usr/share/cinnamon/js/misc/util.js:617:29
 ```

**Notify author of applet**
@NicoG60 

**Issue**

Hi, the applet fails to load when timeshift is configured to use a different mounted partition for snapshots.

`steve@steve-Inspiron-5580:~$ sudo timeshift --list
Device : /dev/nvme0n1p6
UUID   : 398e2ad0-3f28-414f-a71e-ea46c9b8a3a0
Path   : /mnt/virtualbox_images
Mode   : RSYNC
Status : OK
3 snapshots, 46.7 GB free

Num     Name                 Tags  Description  
------------------------------------------------------------------------------
0    >  2020-04-28_05-00-01  M                  
1    >  2020-05-01_06-00-01  D                  
2    >  2020-05-02_06-00-01  D W  `

**Comment by smurphos** [comment]
I can get the applet to load by amending line 126 from

`if(root_uuid =this.backup_device_uuid)`
to
`if(root_uuid === this.backup_device_uuid)`

But it's still not finding the correct mount point reporting 'Backup device not found'

**Comment by smurphos** [comment]
In shell scripting I could get the mountpoint like this..I wonder if that would be a better 'horrible hack' ;-)

`lsblk -o UUID,MOUNTPOINT|grep "$uuid"|cut -d" " -f2`

**Comment by smurphos** [comment]
Hi, 

I've got it pretty much working with a new check_backup_device() function

`    check_backup_device() {
        let [success, result] = GLib.spawn_command_line_sync('bash -c "lsblk -o UUID,MOUNTPOINT | grep ' + this.backup_device_uuid + '"')
        let result_str = result.toString()
        let [uuid, mount] = result_str.split(" ")
        this.backup_mount_point = Gio.File.new_for_path(mount.trim())
        if(success)
            this.set_state(WAIT_FOR_SNAP)
        else
            this.set_state(IDLE)
    }`

**Comment by smurphos** [comment]
Another minor problem - the Create snapshot function fails to launch the polkit dialog with xsession errors getting the dlightfully worded _Refusing to render service to dead parents_ error meesage.

Easily fixed via

```
    make_snapshot() {
    	GLib.spawn_command_line_async('sh -c "pkexec timeshift --create"')
    }
```

**Comment by smurphos** [comment]
Hi, me again - so I've been through the applet and fixed a couple of other minor issues - the wrong tooltip was being displayed in the Wait for Snap state for for example. I think it's all working OK on my set up but need to double-check that the applet picks up on a scheduled snapshot.

@NicoG60 - I'm happy to PR these changes but would you like to test your end as-well?  I've put the modified applet.js here in full temporarily - https://gist.github.com/smurphos/a1484aba428bbffd2c8d76a30c251bd1

**Comment by NicoG60** [comment]
Hey man,
Cheers for the code, That's really good, Thanks a lot! I'll integrate that in my next PR.
I have run across another issue that I'd like to fix and do only one PR for the two.

Maybe you can help me. If you want to dig into it. this issue is:
When the backup device is connected but not mounted, starting timeshift (either gtk or command line) mount it automatically. So far it's fine, but since timeshift is executed as root, the device is mounted in /media/root/<device_name>. Since the applet runs as the user, it does not have the permission to read in that directory which make the applet to crash.
My quick fix was to simply catch the exception and make the error appear in the icon / tooltip. I can't think of any way to work around that, if you have any idea please shout!

**Comment by NicoG60** [comment]
Fixed this issue in PR #3002
Still have the permission issue though... Open to discussion

**Comment by smurphos** [comment]
No worries you are welcome - Unfortunately I've come across another non-working setup with the revised code - this one with timeshift  saving to a SD card mounted at /mnt/external. Difference here is that this machine both the internal drive and the SD card are LUKS encrypted and this seems to confuse Gio.

**Comment by NicoG60** [comment]
Hmm, Actually I revised a bit your code to fit with the recent changes I was making. Try with this version?
But If it is a Gio issue with accessing the actual drive, I don't think there is much we can do.
The applets already feels a bit like bodge as there no proper way to detect timeshift snapshot, adding layers and layers (like Gio binging + LUKs + whatever)  of stuff on top of that won't make it work better :/

**Comment by smurphos** [comment]
Thanks - the new version loads fine on the 2nd machine but fails to find the mount point.
I've worked out why - on that machine, with one longer than standard UUIDs (I guess due to the LUKS encryption) the lsblk output has more than one space between UUID and MOUNTPOINT so 'let [uuid, mount] = result_str.split(" ")' fails to assign anything to mount.

However using a regular expression instead of a space works fine 'let [uuid, mount] = result_str.split(/[ ]+/)'

**Comment by smurphos** [comment]
One other thing to consider - the quick check for a open timeshift-gtk window I added relies on wmctrl being installed on the system, which isn't guaranteed albeit it is a default app for Mint Cinnamon. I'm sure there is a way to this in pure cjs...

**Comment by smurphos** [comment]
Hi - so whilst looking for workarounds to the backup location being /media/root/.... 

I came across another issue. `let [uuid, mount] = result_str.split(/[ ]+/)` fails to get a valid mount if the path includes spaces.
This alternative construction to strip the UUID from the string seems to work OK when the path includes a space.
`let mount = result_str.replace(this.backup_device_uuid,"")`

Re the /media/root issue - a one of `sudo chmod 751 /media/root` on the part of the user - fixes the problem, so maybe the tooltip could reflect that?

**Comment by NicoG60** [comment]
for the `let [mount, uid]` part, what about using `findmnt -o UUID,TARGET -J` which output a json tree with all the mounts on the system, I can filter the results from there. It is a bit of overhead as it would need to parse the output and traverse the tree but at least the space issue should be gone.

for the `chmod` thing, I thought about it and I don't want to introduce security vulnerability giving a permission javascript shouldn't have. It is true that it is the user that run the command and "authorise" that access but I don't want newbies to blindly run a sudo command without understanding what it means. Maybe I'm overthinking...

**Comment by smurphos** [comment]
Yeah that should work although I'm pretty convinced the existing method but using using the .replace method to filter out the UUID is robust, I did double check the syntax and behaviour of .replace() and it does only act on the first instance found in the string. 

Re the chmod issue - the default permission are 750. I can't imagine that adding the executable bit only for all users creates any security issues. File content still can't be read or written to, but the executable bit allows for the presence of the files to be detected.

Any thoughts on additional features? One thing that I think would be very helpful is an alert / warning if the user has the default / as the snapshot location and space is a bit limited. The algorithm to work that out could be quite complex but I think a quick rule of thumb check could be sufficient. E.g. If freespace on the partition is less that the space used by the files that are candidates to be backed up by timeshift then warn the user.

**Comment by NicoG60** [comment]
Yeah, then the `replace()` method is probably more efficient. For the chmod thing, That's true, I didn't realise it was only the executable bit that was added, I saw `chmod` and my brain went "that does not feel right".

For the freespace feature, that's a cool thing! Although I don't really see how to get the next snapshot size. it could be super tiny if nothing really changed or super big if the user added its home directory and have 4 new movies to backup.
